domain/port/state/state.go

```diff
@@ -0,0 +1,271 @@
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return nil
+		}
+		return domain.CoerceError(err)
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	groupedPortRanges := network.GroupedPortRanges{}
+	for _, endpointPortRange := range results {
+		if _, ok := groupedPortRanges[endpointPortRange.Endpoint]; !ok {
+			groupedPortRanges[endpointPortRange.Endpoint] = []network.PortRange{}
+		}
+		groupedPortRanges[endpointPortRange.Endpoint] = append(groupedPortRanges[endpointPortRange.Endpoint], decodeEndpointPortRange(endpointPortRange))
+	}
+	return groupedPortRanges, nil
+}
+
+// getOpenedPorts returns the opened ports for a given unit's endpoint.
+func (st *State) getOpenedPorts(ctx context.Context, tx *sqlair.TX, unitEndpointUUID unitEndpointUUID) ([]portRange, error) {
+	var result []portRange
+
+	selectOpenedPorts, err := st.Prepare(`
+SELECT &portRange.*
+FROM port_range
+JOIN protocol ON port_range.protocol_id = protocol.id
+WHERE unit_endpoint_uuid = $unitEndpointUUID.uuid
+`, portRange{}, unitEndpointUUID)
+
+	if err != nil {
+		return nil, errors.Annotate(err, "preparing select opened ports statement")
+	}
+
+	err = tx.Query(ctx, selectOpenedPorts, unitEndpointUUID).GetAll(&result)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		return []portRange{}, nil
+	}
+	return result, domain.CoerceError(err)
+}
+
+// UpdateUnitPorts opens and closes ports for a given unit. Ports are grouped by endpoint.
+//
+// NOTE: This method opens ports first, and then closes ports. However, it is recommended
+// that your service layer reconciles overlapping port ranges before calling this method.
+func (st *State) UpdateUnitPorts(ctx context.Context, unitUUID string, openPorts, closePorts network.GroupedPortRanges) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	endpointsUnderAction := set.NewStrings()
+	for endpoint := range openPorts {
+		endpointsUnderAction.Add(endpoint)
+	}
+	for endpoint := range closePorts {
+		endpointsUnderAction.Add(endpoint)
+	}
+
+	getProtocolID, err := st.Prepare("SELECT &protocolID.* FROM protocol WHERE protocol = $protocolName.protocol", protocolID{}, protocolName{})
+	if err != nil {
+		return errors.Annotate(err, "preparing get protocol ID statement")
+	}
+
+	insertPortRange, err := st.Prepare("INSERT INTO port_range (*) VALUES ($unitPortRange.*)", unitPortRange{})
+	if err != nil {
+		return errors.Annotate(err, "preparing insert port range statement")
+	}
+
+	clearPortRange, err := st.Prepare(`
+DELETE FROM port_range
+WHERE unit_endpoint_uuid = $unitEndpointUUID.uuid
+`, unitEndpointUUID{})
+	if err != nil {
+		return errors.Annotate(err, "preparing clear port range statement")
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+
+		unitEndpointMap := map[string]unitEndpointUUID{}
+		for _, endpoint := range endpointsUnderAction.Values() {
+			var err error
+			unitEndpointMap[endpoint], err = st.getUnitEndpointUUID(ctx, tx, unitUUID, endpoint)
```

Comment:
It feels like having queries in a for loop like this is a bit of an antipattern, each of these requires a network request to the DB, could they not be done in a GetAll and bulk insert instead?

Reply:
The processesing of the arguemnts to the new queries could then be done outside of the Txn too

Reply:
Good idea. I originally wrote this forgetting it would need to be wrapped in a loop. You're right this should be take out.

Reply:
This has been through a major refactor since you looked.